### PR TITLE
FB8-251: Add ability to build with BoringSSL

### PR DESCRIPTION
**Pull request by Sudokamikaze**

### CODE REVIEW

File: local/build-binary [hunk 1]
```diff
@@ -61,6 +62,29 @@ fi
...
 
+if [[ ${WITH_BORINGSSL} == "true" ]]; then
+    # ------------------------------------------------------------------------------
+    # compile boringssl libs
+    # ------------------------------------------------------------------------------
+    git clone https://github.com/google/boringssl.git
+    mkdir boringssl/bin
+    cd boringssl/bin
+    if [[ -f /etc/redhat-release ]]; then
+        source /opt/rh/go-toolset-7/enable
```
Comment (EvgeniyPatlan) [inline local/build-binary]: go-toolset is installed for rhel7. So on centos6 and centos8 there will be an issue

**Review by EvgeniyPatlan — CHANGES_REQUESTED** [review]

File: local/build-binary (hunk 1 above)
Comment (Sudokamikaze) [reply local/build-binary]: Added support for RHEL8 and added missing package for unwind tests

And added a check for centos 6, as BoringSSL can't compile on it

File: docker/install-deps [hunk 2]
```diff
@@ -61,7 +61,11 @@ if [ -f /usr/bin/yum ]; then
     "
 
     if [[ ${RHVER} -eq 8 ]]; then
-        PKGLIST+=" lz4 perl-Memoize perl-open valgrind-devel libubsan rpcgen libtirpc-devel"
+        PKGLIST+=" lz4 perl-Memoize perl-open valgrind-devel libubsan rpcgen libtirpc-devel go libunwind-devel"
```
Comment (EvgeniyPatlan) [inline docker/install-deps]:
```suggestion
        PKGLIST+=" lz4 perl-Memoize perl-open valgrind-devel libubsan rpcgen libtirpc-devel golang libunwind-devel"
```

**Review by EvgeniyPatlan — CHANGES_REQUESTED** [review]

**Comment by Sudokamikaze** [comment]
Link to test job: https://fb.cd.percona.com/job/fb-mysql-server-8.0-pipeline-illiapshonkin/

File: docker/prepare-docker [hunk 3]
```diff
@@ -15,7 +15,7 @@ build_docker() {
     docker build \
         --squash \
         --no-cache \
-        -t perconalab/ps-build:${SOURCE_IMAGE//[:\/]/-} \
```
Comment (hors) [inline docker/prepare-docker]: @Sudokamikaze we will use 'ps-build' not 'fb-*'

Comment (Sudokamikaze) [reply docker/prepare-docker]: Removed deps install and rebased them to jenkins-pipeline repo

**Comment by Sudokamikaze** [comment]
Link to PS build with changeset: https://ps3.cd.percona.com/view/8.0/job/percona-server-8.0-pipeline/3605/
Link to FB build: https://fb.cd.percona.com/job/fb-mysql-server-8.0-pipeline-illiapshonkin/
Link to FB build without changes: https://fb.cd.percona.com/view/8.0-fb/job/fb-mysql-server-8.0-pipeline/

@hors @EvgeniyPatlan @percona-ysorokin

**Review by EvgeniyPatlan — APPROVED** [review]

**Review by hors — APPROVED** [review]

File: local/build-binary [hunk 4]
```diff
@@ -61,6 +62,38 @@ fi
...
+                source /opt/rh/devtoolset-8/enable
+                BORING_SSL_CMAKE=/usr/bin/cmake3
+            ;;
+            6) echo "Unsupported CentOS version" && exit 1 ;;
+        esac
+    else
+        BORING_SSL_CMAKE=/usr/bin/cmake
+    fi
+    ${BORING_SSL_CMAKE} .. -DCMAKE_CXX_FLAGS="-fPIC" -DCMAKE_C_FLAGS="-fPIC"
+    if make -j$(nproc); then
```
Comment (percona-ysorokin) [inline local/build-binary:89]: Should we fail if there was a problem building BoringSSL?

File: local/build-binary [hunk 5]
```diff
@@ -61,6 +62,38 @@ fi
...
+                fi
+                source /opt/rh/devtoolset-8/enable
+                BORING_SSL_CMAKE=/usr/bin/cmake3
+            ;;
+            6) echo "Unsupported CentOS version" && exit 1 ;;
+        esac
+    else
+        BORING_SSL_CMAKE=/usr/bin/cmake
+    fi
+    ${BORING_SSL_CMAKE} .. -DCMAKE_CXX_FLAGS="-fPIC" -DCMAKE_C_FLAGS="-fPIC"
```
Comment (percona-ysorokin) [inline local/build-binary:88]: Please, make sure that directly assigning `CMAKE_CXX_FLAGS` and `CMAKE_C_FLAGS` does not overwrite other flags.
To be honest well-written CMake should handle this case properly, but MySQL/Percona Server, for instance, does not.

File: local/build-binary (hunk 4 above)
Comment (Sudokamikaze) [reply local/build-binary:89]: Yes it should fail in case of compilation failure, I'll add `else; exit 1`

File: local/build-binary (hunk 5 above)
Comment (Sudokamikaze) [reply local/build-binary:88]: Would such construction be sufficient for this purpose?
```
    BORING_SSL_FLAGS="-fPIC"
    if ${BORING_SSL_CMAKE} .. -DCMAKE_CXX_FLAGS="${BORING_SSL_FLAGS}" -DCMAKE_C_FLAGS="${BORING_SSL_FLAGS}"; then
        unset BORING_SSL_FLAGS
    fi
```

Comment (percona-ysorokin) [reply local/build-binary:88]: I double checked with the BoringSSL source code.
Here
https://github.com/google/boringssl/blob/master/CMakeLists.txt#L153
I believe everything is fine, according to this line final `CMAKE_CXX_FLAGS` is formed based on the value passed via command line options. Only then custom compiler options from the cmake file are appended.
So, no changes required here. Initial `${BORING_SSL_CMAKE} .. -DCMAKE_CXX_FLAGS="-fPIC" -DCMAKE_C_FLAGS="-fPIC"` is fine.

Comment (Sudokamikaze) [reply local/build-binary:88]: Understood, thank you

File: local/build-binary (hunk 4 above)
Comment (Sudokamikaze) [reply local/build-binary:89]: re: Added in 7d02b5e

**Comment by Sudokamikaze** [comment]
Here's a working build with BoringSSL changeset: https://fb.cd.percona.com/job/fb-mysql-server-8.0-boringssl-param/9/
(properly show MTRS)

**Review by percona-ysorokin — APPROVED** [review]
LGTM

**Review by inikep — COMMENTED** [review]
LGTM

**Pull request merged**